### ISSUE TITLE
cannot execute SELECT FOR NO KEY UPDATE in a read-only transaction

### ISSUE THREAD

**Issue by kwongtn98** [postgresml/pgcat#887]
**Describe the bug**
We have the following sql statement
```sql
SELECT mytable.id, mytable.other_id
FROM mytable
WHERE (mytable.status = %s
       OR (mytable.counter< (mytable.max_counter)
           AND mytable.next_attempt_at <= (STATEMENT_TIMESTAMP())
           AND mytable.status = %s))
ORDER BY mytable.next_attempt_at ASC NULLS FIRST,
                                                                   mytable.created ASC
LIMIT 10
FOR NO KEY
UPDATE OF mytable SKIP LOCKED

```

Our django application then throws
```
cannot execute SELECT FOR NO KEY UPDATE in a read-only transaction
```

From a high level view it seems like an error with the query parser.

**Comment by gcsabbagh** [comment]
Yes, there is a bug in the role selection logic, and it seems to default back to "auto" or "any" if query parser fails.

I opened an issue here: https://github.com/postgresml/pgcat/issues/884

Until it's fixed, in order to avoid this, your options are:
- Turn off query parser
- Or `SET SERVER ROLE TO 'primary'` in your transaction/session

**Comment by sriram-sundar-41** [comment]
@gcsabbagh @kwongtn98 Has this issue been fixed? I see the query parser fails for the below config and the query still goes to read replicas and fails. Any pointers are appreciated

`
2025-05-06T18:41:41.991395Z DEBUG ThreadId(04) pgcat::query_router: sql parser error: Expected an SQL statement, found: DO at Line: 2, Column 1: SET client_min_messages = NOTICE; SET statement_timeout = '600000'; 
2025-05-06T18:41:41.991401Z  WARN ThreadId(04) pgcat::client: Query parsing error: QueryRouterParserError("sql parser error: Expected an SQL statement, found: DO at Line: 2, Column 1") (client: { application_name: psql, username: username, pool_name: vcdb })
`

>     [general]
>     admin_username = "pgcat_admin"
>     admin_password = "pgcat_password"
>     port = 6432
>     connect_timeout = 10000
>     idle_timeout = 300000
>     shutdown_timeout = 30000
>     healthcheck_delay = 5000
>     healthcheck_timeout = 2000
>     server_tls = true
>     verify_server_certificate = false
>     tls_certificate = "/etc/pgcat/tls/tls.crt"
>     tls_private_key = "/etc/pgcat/tls/tls.key"
>     log_client_connections = true
>     log_client_disconnections = true
> 
>     [pools.vcdb]
>     prepared_statements = true
>     prepared_statements_cache_size = 50000
>     query_parser_enabled = true
>     query_parser_read_write_splitting = true
>     default_role = "primary"
> 
>     [pools.vcdb.users.0]
>     username = "username"
>     password = "password"
>     pool_mode = "transaction"
>     pool_size = 100
>     min_pool_size = 10
> 
>     [pools.vcdb.shards.0]
>     database = "vcdb"
>     servers = [
>       { host = "postgres-instance.vidb-external", port = 5432, role = "primary" },
>       { host = "postgres-instance-repl.vidb-external", port = 5432, role = "replica" },
>     ]
>

**Comment by gcsabbagh** [comment]
No one is actively working on pgcat. Main creator is now working on a successor: https://github.com/pgdogdev/pgdog

I will migrate my apps to that once it's stable
